File: backend/blackbox/src/main/java/com/insight_web/blackbox/usecase/UserRegistration.java
package com.insight_web.blackbox.usecase;

import com.insight_web.blackbox.domain.Role;
import com.insight_web.blackbox.domain.User;
import com.insight_web.blackbox.domain.exception.ResourceNotFoundException;
import com.insight_web.blackbox.usecase.action.RegisterAction;
import com.insight_web.blackbox.usecase.storage.UserStorage;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;

@Service
@RequiredArgsConstructor
public class UserRegistration {
    private final UserStorage userStorage;

    @Transactional
    public void register(RegisterAction action) {
        if (!action.getPassword().equals(action.getPasswordConfirmation())) {
            throw new RuntimeException("Password and password does not match");
        }
        if (!userStorage.existsByEmail(action.getEmail())) {
            throw new ResourceNotFoundException("user by email " + action.getEmail() + "does not found");
        }
        var user = action.toDomainModel();

        userStorage.create(user);
    }
}

File: backend/auth/src/main/kotlin/com/insightweb/auth/usecase/UserRegistration.kt
package com.insightweb.auth.usecase

import com.insightweb.auth.domain.User
import com.insightweb.auth.domain.exception.ResourceNotFoundException
import com.insightweb.auth.usecase.action.RegisterAction
import com.insightweb.auth.usecase.storage.UserStorage
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class UserRegistration(
    private val userStorage: UserStorage,
) {

    @Transactional
    fun register(action: RegisterAction) {
        if (action.password != action.passwordConfirmation) {
            throw RuntimeException("Password and password does not match")
        }
        if (!userStorage.existsByEmail(action.login)) {
            throw ResourceNotFoundException("user by email " + action.login + "does not found")
        }
        val user: User = action.toUser()
        userStorage.create(user)
    }
}